### PR TITLE
[sns] added possibility to define already existing topics (prevent create topic call) #1022

### PR DESCRIPTION
**Pull request by paramonov**

### CODE REVIEW

File: pkg/sns/SnsContext.php [hunk 1]
```diff
@@ -31,12 +31,11 @@ class SnsContext implements Context
...
     private $topicArns;
 
-    public function __construct(SnsClient $client, array $config)
+    public function __construct(SnsClient $client, array $config, array $topicArns = [])
     {
         $this->client = $client;
         $this->config = $config;
-
-        $this->topicArns = [];
+        $this->topicArns = $topicArns;
```
Comment (makasim) [inline pkg/sns/SnsContext.php:38]: topicArns property format is 

```
topic_name => arn 
```

File: pkg/sns/SnsConnectionFactory.php [hunk 2]
```diff
@@ -71,7 +72,7 @@ public function __construct($config = 'sns:')
      */
     public function createContext(): Context
     {
-        return new SnsContext($this->establishConnection(), $this->config);
+        return new SnsContext($this->establishConnection(), $this->config, $this->config['topic_arns']);
```
Comment (makasim) [inline pkg/sns/SnsConnectionFactory.php:75]: why do you pass topc_arns as separate arg? They are already there in config which is passed as second arg

Comment (paramonov) [reply pkg/sns/SnsConnectionFactory.php]: removed third arg

File: pkg/sns/SnsContext.php (hunk 1 above)
Comment (paramonov) [reply pkg/sns/SnsContext.php]: yep, refactored

File: pkg/sns/SnsConnectionFactory.php [hunk 3]
```diff
@@ -59,6 +60,9 @@ public function __construct($config = 'sns:')
 
                 unset($config['dsn']);
             }
+            if (\array_key_exists('topic_arns', $config) && \is_string($config['topic_arns'])) {
```
Comment (makasim) [inline pkg/sns/SnsConnectionFactory.php:63]: at this stage it has to be assoc array already. The string parsing should happen if dsn is passed.

File: pkg/sns/Tests/SnsConnectionFactoryConfigTest.php [hunk 4]
```diff
@@ -148,6 +154,24 @@ public static function provideConfigs()
                 'version' => '2010-03-31',
                 'lazy' => false,
                 'endpoint' => 'http://localstack:1111',
+                'topic_arns' => [],
+            ],
+        ];
+
+        yield [
+            ['dsn' => 'sns:?topic_arns=topic1|arn:aws:sns:us-east-1:123456789012:topic1;topic2|arn:aws:sns:us-west-2:123456789012:topic2'],
```
Comment (makasim) [inline pkg/sns/Tests/SnsConnectionFactoryConfigTest.php:162]: what about passing arns as

```
topic_arns[topic1]=arn:aws:sns:us-east-1:123456789012&topic_arns[topic2]=arn:aws:sns:us-west-2:123456789012:topic2
```

that way the standard query parsing function does the job

Comment (paramonov) [reply pkg/sns/Tests/SnsConnectionFactoryConfigTest.php]: good point, replaced

File: pkg/sns/SnsConnectionFactory.php (hunk 3 above)
Comment (paramonov) [reply pkg/sns/SnsConnectionFactory.php:63]: removed, not needed when using dsn parsing

File: pkg/sns/SnsContext.php [hunk 5]
```diff
@@ -35,8 +35,7 @@ public function __construct(SnsClient $client, array $config)
     {
         $this->client = $client;
         $this->config = $config;
-
-        $this->topicArns = [];
+        $this->topicArns = $config['topic_arns'];
```
Comment (makasim) [inline pkg/sns/SnsContext.php:38]: this is BC break. set empty array if not set

**Comment by makasim** [comment]
fixes #1022

**Pull request merged**